File: MssgTCP.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.Socket;
import java.nio.channels.ServerSocketChannel;

public class MssgTCP {

	protected static Entite insertNouveauTCP(Entite entite, boolean joindre) {
		try {
			Socket socket_tcp = new Socket(entite.getAddrNext(1), entite.getPortTCPOut());
			BufferedReader br = new BufferedReader(new InputStreamReader(socket_tcp.getInputStream()));
			PrintWriter tcp_pw = new PrintWriter(new OutputStreamWriter(socket_tcp.getOutputStream()));

			// WELC ou NOTC
			String message = receiveTCP(br, entite, tcp_pw, socket_tcp);
			analyseMssg(message);
			message = Annexe.substringLast(message);

			String[] parts = message.split(" ");
			if (parts[0].equals("WELC")) {
				if (joindre == true) {
					entite.setAddrNext(parts[1], 1);
					entite.setPortOutUDP(Integer.parseInt(parts[2]), 1);
					entite.setAddrMultiDiff(parts[3], 1);
					entite.setPortMultiDiff(Integer.parseInt(parts[4]), 1);
				}

				String send = "";
				if (joindre == true) {
					send = "NEWC " + Annexe.trouveAdress(true) + " " + entite.getPortInUDP() + "\n";
				} else {
					send = "DUPL " + Annexe.trouveAdress(true) + " " + entite.getPortInUDP() + " "
							+ entite.getAddrMultiDiff(1) + " " + entite.getPortMultiDiff(1) + "\n";
				}
				analyseMssg(send);
				sendTCP(send, tcp_pw);

				// ACKC ou ACKD
				message = receiveTCP(br, entite, tcp_pw, socket_tcp);
				analyseMssg(message);
				message = Annexe.substringLast(message);
				if (joindre == false) {
					parts = message.split(" ");
					entite.setPortOutUDP(Integer.parseInt(parts[1]), 1);
				}
				closeTCP(entite, br, tcp_pw, socket_tcp);
				print(false, entite);
			} else {
				System.out.println("Entite ou se connecter est deja un doubleur");
				closeTCP(entite, br, tcp_pw, socket_tcp);
				print(true, entite);
				System.exit(0);
			}

		} catch (IOException e) {
			e.printStackTrace();
		} catch (MssgSpellCheck e) {
			e.getMessage();
		} catch (LengthException e) {
			e.getMessage();
			System.exit(0);
		} catch (NotSDLException e) {
			e.getMessage();
			System.exit(0);
		}
		return entite;
	}

	protected static Entite insertAnneauTCP(Entite entite, ServerSocketChannel tcp_in_ssc) {
		if (Main.affichage) {
			System.out.println("Evenement sur TCP");
		}
		try {
			Socket sock_tcp = tcp_in_ssc.socket().accept();
			if (Main.affichage) {
				System.out.println("Acceptation TCP");
			}
			BufferedReader tcp_br = new BufferedReader(new InputStreamReader(sock_tcp.getInputStream()));
			PrintWriter tcp_pw = new PrintWriter(new OutputStreamWriter(sock_tcp.getOutputStream()));

			String mssg = "";
			if (entite.getIsDuplicateur() == false) {
				// WELC
				mssg = "WELC " + entite.getAddrNext(1) + " " + entite.getPortOutUDP(1) + " "
						+ entite.getAddrMultiDiff(1) + " " + entite.getPortMultiDiff(1) + "\n";
			} else {
				// NOTC
				mssg = "NOTC\n";
			}
			analyseMssg(mssg);
			sendTCP(mssg, tcp_pw);

			if (entite.getIsDuplicateur() == false) {
				// NEWC ou DUPL
				String lu = receiveTCP(tcp_br, entite, tcp_pw, sock_tcp);
				analyseMssg(lu);
				lu = Annexe.substringLast(lu);
				String parts[] = lu.split(" ");
				String futurAddrUDPOut = parts[1];
				String futurPortUDPOut = parts[2];
				int anneau = 1;
				String futurMultiDiffAddr = "";
				String futurMultiDiffPort = "";
				String envoi = "";
				boolean demandeDupplication = false;
				if (parts[0].equals("NEWC")) {
					envoi = "ACKC\n";
				} else if (parts[0].equals("DUPL")) {
					anneau = 2;
					futurMultiDiffAddr = parts[3];
					futurMultiDiffPort = parts[4];
					envoi = "ACKD " + entite.getPortInUDP() + "\n";
					demandeDupplication = true;
				}
				analyseMssg(envoi);
				sendTCP(envoi, tcp_pw);

				entite.setAddrNext(futurAddrUDPOut, anneau);
				entite.setPortOutUDP(Integer.parseInt(futurPortUDPOut), anneau);
				if (parts[0].equals("DUPL")) {
					entite.setAddrMultiDiff(futurMultiDiffAddr, 2);
					entite.setPortMultiDiff(Integer.parseInt(futurMultiDiffPort), 2);
				}

				closeTCP(entite, tcp_br, tcp_pw, sock_tcp);
				entite.setIsDuplicateur(demandeDupplication);
				print(false, entite);
			} else {
				closeTCP(entite, tcp_br, tcp_pw, sock_tcp);
				print(true, entite);
			}
		} catch (IOException e) {
			e.printStackTrace();
		} catch (MssgSpellCheck e) {
			e.getMessage();
		} catch (LengthException e) {
			e.getMessage();
		} catch (NotSDLException e) {
			e.getMessage();
		}
		return entite;
	}

	private static void print(boolean erreur, Entite entite) {
		if (Main.affichage) {
			if (erreur) {
				System.out.println("Erreur lors de l'insertion donc pas d'insertion -- entite inchange\n");
			} else {
				entite.printEntiteSimple();
			}
		}
	}

	private static void closeTCP(Entite entite, BufferedReader tcp_br, PrintWriter tcp_pw, Socket sock_tcp) {
		try {
			tcp_br.close();
			tcp_pw.close();
			sock_tcp.close();
			if (Main.affichage) {
				System.out.println("Fin de connection TCP");
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private static void sendTCP(String envoi, PrintWriter tcp_pw) {
		tcp_pw.print(envoi);
		tcp_pw.flush();
		if (Main.affichage) {
			System.out.println("Envoi de : " + envoi);
		}
	}

	private static String receiveTCP(BufferedReader br, Entite entite, PrintWriter tcp_pw, Socket sc) {
		try {

			char[] data = new char[512];
			int mssg;
			mssg = br.read(data);
			boolean erreur = false;
			if (mssg != -1) {
				String message = new String(data, 0, mssg);
				if (Main.affichage) {
					System.out.println("Recu de : " + message);
				}
				return message;
			} else {
				erreur = true;
				closeTCP(entite, br, tcp_pw, sc);
				print(erreur, entite);
			}
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(0);
		}
		return null;
	}

	private static void analyseMssg(String mssg) throws LengthException, NotSDLException, MssgSpellCheck {
		String parts[] = mssg.split(" ");
		if (mssg.length() > Main.SIZEMESSG) {
			throw new LengthException(mssg.length(), mssg, "TCP");
		}
		if (parts[0].equals("WELC") || parts[0].equals("DUPL")) {
			suiteAnalyseMssg(5, mssg, parts);
		} else if (parts[0].equals("ACKC\n") || parts[0].equals("NOTC\n")) {
			suiteAnalyseMssg(1, mssg, parts);
		} else if (parts[0].equals("ACKD")) {
			suiteAnalyseMssg(2, mssg, parts);
		} else if (parts[0].equals("NEWC")) {
			suiteAnalyseMssg(3, mssg, parts);
		} else {
			throw new MssgSpellCheck(parts[0], "TCP");
		}
	}

	private static void suiteAnalyseMssg(int longeur, String mssg, String[] parts)
			throws LengthException, NotSDLException {
		if (parts.length != longeur) {
			throw new LengthException(longeur, parts.length, "TCP", mssg);
		}
		if (!(parts[longeur - 1].substring(parts[longeur - 1].length() - 1, parts[longeur - 1].length())
				.equals("\n"))) {
			throw new NotSDLException(mssg, "TCP");
		}
	}
}

File: C/MssgTCP.c
#include "MssgTCP.h"

void sendTCP(char* mess, int sock){
  write(sock, mess, strlen(mess));
  char tmp[SIZEMSSG+11];
  sprintf(tmp, "%s %s","Envoi de :",mess);
  INFO(tmp);
}

char* receiveTCP(int sock){
  char* buff= malloc(SIZEMSSG * sizeof(char));
  if(buff == NULL ){
    fprintf(stderr,"Allocation impossible : %s\n","fonction receiveTCP : MssgTCP.c");
    return NULL;
  }
  int recu=read(sock, buff, (SIZEMSSG-1) * sizeof(char));
  if(recu >= 0){
    buff[recu] = '\0';
    char tmp[SIZEMSSG+14];
    sprintf(tmp, "%s %s","Message recu :",buff);
    INFO(tmp);
  }
  return buff;
}

bool suiteAnalyseMssg(int longeur, char* mssg, char** parts, int sizeParts){
  bool res = false;
  if (sizeParts != longeur) {
    lengthException(longeur, sizeParts, "TCP", mssg);
    res = true;
  }
  int sizeMssg = strlen(parts[sizeParts-1]);
  char lastMssg = parts[sizeParts-1][sizeMssg-1];
  if (lastMssg != '\n') {
    notSDLException(mssg, "TCP");
    res = true;
  }
  return res;
}

bool analyseMssg(char * mssg, char** parts, int nombreDePartie ) {
  bool res = false;
  int sizeMssg = strlen(mssg);
  if (sizeMssg > SIZEMSSG) {
    lengthExceptionMax(sizeMssg, mssg, "TCP");
    res = true;
  }
  if (strcmp(parts[0],"WELC")==0 || strcmp(parts[0],"DUPL")==0) {
    res = suiteAnalyseMssg(5, mssg, parts,nombreDePartie);
  } else if (strcmp(parts[0],"ACKC\n")==0 || strcmp(parts[0],"NOTC\n")==0) {
    res = suiteAnalyseMssg(1, mssg, parts,nombreDePartie);
  } else if (strcmp(parts[0],"ACKD")==0) {
    res = suiteAnalyseMssg(2, mssg, parts,nombreDePartie);
  } else if (strcmp(parts[0],"NEWC")==0) {
    res = suiteAnalyseMssg(3, mssg, parts,nombreDePartie);
  } else {
    mssgSpellCheck(parts[0], "TCP");
    res = true;
  }
  return res;
}

void print(bool erreur, Entite* entite) {
  if (erreur == true) {
    INFO("Erreur lors de l'insertion donc pas d'insertion -- entite inchange\n");
  } else {
    INFO(entite_print_simple(entite));
  }
}

Entite* insertNouveauTCP(Entite* entite, bool joindre) {
  bool res = false;
  struct sockaddr_in *adress_sock;
  struct addrinfo *first_info;
  struct addrinfo hints;
  bzero(&hints,sizeof(struct addrinfo));
  hints.ai_family = AF_INET;
  hints.ai_socktype = SOCK_STREAM;
  int r2 = getaddrinfo(convertIPV4Imcomplete(getAddrNext(entite,1)),getPortTCPOut(entite),&hints,&first_info);
  if(r2 == -1){
    fprintf(stderr,"pb InsertNOUVEAU TCP\n");
  }
  int sock = socket(PF_INET, SOCK_STREAM, 0);
  if(first_info != NULL){
    adress_sock = (struct sockaddr_in *)first_info->ai_addr;
    int r = connect(sock,(struct sockaddr *)adress_sock, (socklen_t)sizeof(struct sockaddr_in));
    if(r != 1){
      char * message = receiveTCP(sock);
      char* copyMessage = copyStr(message);
      char* copyMessage2 = copyStr(message);
      int nombreDePartie = sizePart(copyMessage, " ");
      char** parts = split(message, ' ');
      res = analyseMssg(copyMessage2, parts, nombreDePartie);
      parts = substringLast(parts, nombreDePartie);
      if(res == true){
        close(sock);
        exit(EXIT_FAILURE);
      }
      if (strcmp(parts[0], "WELC")==0) {
        if (joindre == true) {
          entite = setAddrNext(entite, parts[1], 1);
          entite = setPortOutUDP(entite, parts[2], 1);
          entite = setAddrMultiDiff(entite, parts[3], 1);
          entite = setPortMultiDiff(entite, parts[4], 1);
        }
        char* send = malloc(SIZEMSSG*sizeof(char));
        if(send == NULL ){
          fprintf(stderr,"Allocation impossible : %s\n","fonction insertNouveauTCP : MssgTCP.c");
          return NULL;
        }
        if (joindre == true) {
          sprintf(send, "%s%s %s\n", "NEWC ", trouveAdress(true), getPortInUDP(entite));
        } else {
          sprintf(send, "%s%s %s %s %s\n", "DUPL " , trouveAdress(true) , getPortInUDP(entite) , getAddrMultiDiff(entite,1) , getPortMultiDiff(entite, 1));
        }
        copyMessage = copyStr(send);
        copyMessage2 = copyStr(send);
        char* copyMessage3 = copyStr(send);
        nombreDePartie = sizePart(copyMessage, " ");
        parts = split(copyMessage3, ' ');
        analyseMssg(copyMessage2, parts, nombreDePartie);
        sendTCP(send, sock);
        free(send);

        // ACKC ou ACKD
        message = receiveTCP(sock);
        copyMessage = copyStr(message);
        copyMessage2 = copyStr(message);
        nombreDePartie = sizePart(copyMessage, " ");
        parts = split(message, ' ');
        res = analyseMssg(copyMessage2, parts, nombreDePartie);
        parts = substringLast(parts, nombreDePartie);

        if (joindre == false) {
          entite = setPortOutUDP(entite, parts[1], 1);
        }

        close(sock);
        INFO("Fin de connection TCP");
        print(false, entite);
        free(copyMessage3);
      } else {
        printf("Entite ou se connecter est deja un doubleur ");
        close(sock);
        INFO("Fin de connection TCP");
        print(true, entite);
        exit(EXIT_FAILURE);
      }
      free(parts);
      free(message);
      free(copyMessage);
      free(copyMessage2);
    }
  }
  return entite;
}

Entite* insertAnneauTCP(Entite* entite, int sock) {
  INFO("Evenement sur TCP\n");
  int sock2;
  struct sockaddr_in caller;
  socklen_t size = sizeof(caller);
  if ((sock2 = accept(sock,(struct sockaddr *)&caller,&size))==-1) {
    fprintf(stderr,"accept problem");
    close(sock);
    exit(EXIT_FAILURE);
  }
  if(sock2 >= 0){
    INFO("Acceptation TCP\n");
    char* send = malloc(SIZEMSSG*(sizeof(char)));
    if(send == NULL ){
      fprintf(stderr,"Allocation impossible : %s\n","fonction insertAnneauTCP : send : MssgTCP.c");
      return NULL;
    }
    if (getIsDuplicateur(entite) == false) {
      // WELC
      sprintf(send,"%s%s %s %s %s\n","WELC ",trouveAdress(), getPortOutUDP(entite,1), getAddrMultiDiff(entite,1), getPortMultiDiff(entite,1));
    } else {
      // NOTC
      sprintf(send, "%s\n", "NOTC");
    }
    char* copyMessage = copyStr(send);
    char* copyMessage2 = copyStr(send);
    char* copyMessage3 = copyStr(send);
    int nombreDePartie = sizePart(copyMessage, " ");
    char** parts = split(copyMessage3, ' ');
    analyseMssg(copyMessage2, parts, nombreDePartie);
    sendTCP(send, sock2);
    free(send);
    if (getIsDuplicateur(entite) == false) {
      // NEWC ou DUPL
      char* message = receiveTCP(sock2);
      copyMessage = copyStr(message);
      copyMessage2 = copyStr(message);
      nombreDePartie = sizePart(copyMessage, " ");
      parts = split(message, ' ');
      analyseMssg(copyMessage2, parts, nombreDePartie);
      parts = substringLast(parts, nombreDePartie);
      char* futurAddrUDPOut = parts[1];
      char* futurPortUDPOut = parts[2];
      int anneau = 1;
      char* futurMultiDiffAddr = malloc(16*sizeof(char));
      if(futurMultiDiffAddr == NULL ){
        fprintf(stderr,"Allocation impossible : %s\n","fonction insertAnneauTCP : futurMultiDiffAddr : MssgTCP.c");
        return NULL;
      }
      char* futurMultiDiffPort = malloc(5*sizeof(char));
      if(futurMultiDiffPort == NULL ){
        fprintf(stderr,"Allocation impossible : %s\n","fonction insertAnneauTCP : futurMultiDiffPort : MssgTCP.c");
        return NULL;
      }
      char* envoi = malloc(SIZEMSSG*sizeof(char));
      if(envoi == NULL ){
        fprintf(stderr,"Allocation impossible : %s\n","fonction insertAnneauTCP : envoi : MssgTCP.c");
        return NULL;
      }
      bool demandeDupplication = false;
      if (strcmp(parts[0], "NEWC")==0) {
        envoi = "ACKC\n";
      } else if (strcmp(parts[0], "DUPL")==0) {
        anneau = 2;
        futurMultiDiffAddr = parts[3];
        futurMultiDiffPort = parts[4];
        sprintf(envoi, "%s%s\n", "ACKD ", getPortInUDP(entite));
        demandeDupplication = true;
      }
      copyMessage = copyStr(envoi);
      copyMessage2 = copyStr(envoi);
      copyMessage3 = copyStr(envoi);
      nombreDePartie = sizePart(copyMessage, " ");
      parts = split(copyMessage3, ' ');
      analyseMssg(copyMessage2, parts, nombreDePartie);
      sendTCP(envoi, sock2);
      entite = setAddrNext(entite, futurAddrUDPOut, anneau);
      entite = setPortOutUDP(entite, futurPortUDPOut, anneau);
      if (demandeDupplication == true) {
        entite = setAddrMultiDiff(entite, futurMultiDiffAddr, anneau);
        entite = setPortMultiDiff(entite, futurMultiDiffPort, anneau);
      }

      close(sock2);
      INFO("Fin de connection TCP");
      entite = setIsDuplicateur(entite, demandeDupplication);
      print(false, entite);
      free(send);
      free(futurAddrUDPOut);
    } else {
      close(sock2);
      INFO("Fin de connection TCP");
      print(true, entite);
    }
    free(parts);
    free(copyMessage);
    free(copyMessage2);
    free(copyMessage3);
  }
  return entite;
}
